### PR TITLE
KEYCLOAK-15847 Add MapUserProvider

### PR DESCRIPTION
**Pull request by mhajas**

### CODE REVIEW

**Comment by mhajas** [comment]
PIpeline run:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2349/
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/1935/

File: server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java [hunk 1]
```diff
@@ -18,15 +18,52 @@
...
+import org.keycloak.models.KeycloakTransaction;
 import org.keycloak.provider.Provider;
 
 /**
  * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
+ *
+ * Note that invocation of any of {@code onEvent} methods doesn't mean that the event happened at the moment but that it
+ * will happen during the main transaction commit. This means the execution, corresponding to the triggered event,
+ * may end up with a failure. Therefore, each implementation of this interface should consider using an internal
+ * {@link KeycloakTransaction} for firing events only in case the main transaction ends up successfuly. This can be
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java:30]: _Firing_ an event should be precondition for this interface's methods to be invoked. Maybe _persisting_ would express the intention? (same in the javadocs below)

File: server-spi/src/main/java/org/keycloak/models/UserModel.java [hunk 2]
```diff
@@ -56,7 +59,13 @@
     // No default method here to allow Abstract subclasses where the username is provided in a different manner
     String getUsername();
 
-    // No default method here to allow Abstract subclasses where the username is provided in a different manner
+    /**
+     * Sets username for this user, username should be switched to lowercase before setting.
```
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/models/UserModel.java:63]: It should not, there is demand for case-sensitive username support. See e.g. https://issues.redhat.com/browse/KEYCLOAK-5699.
For this moment, Keycloak uses case-insensitive usernames. Is this because of search? Then we need to stress that it needs to be case-insensitive searchable, but that will be done later in the search operation of the underlying storage.

File: server-spi/src/main/java/org/keycloak/models/UserModel.java [hunk 3]
```diff
@@ -129,9 +138,15 @@
 
     void removeRequiredAction(String action);
 
-    void addRequiredAction(RequiredAction action);
+    default void addRequiredAction(RequiredAction action) {
+        String actionName = action.name();
```
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/models/UserModel.java:142]: There should be check for `action != null` (even though I appreciate it's not there in the original JPA variant)

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package org.keycloak.models.map.user;
+
+import java.util.Comparator;
+import java.util.UUID;
+
+public class MapUserEntity extends AbstractUserEntity<UUID> {
+
+    public static final Comparator<MapUserEntity> COMPARE_BY_USERNAME = Comparator.comparing(MapUserEntity::getUsername);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java:25]: Should compare irrespective of case?

**Review by hmlnarik — COMMENTED** [review]
Great job, @mhajas, thank you! I only left few comments in line.

File: testsuite/integration-arquillian/tests/base/src/test/resources/export/partialexport-testrealm.json [hunk 5]
```diff
@@ -1155,6 +1155,9 @@
           ],
           "authType": [
             "simple"
+          ],
+          "enabled": [
+            "false"
```
[the remark this answers is not in the document]
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/resources/export/partialexport-testrealm.json:1160]: Good question, I am not sure it is necessary any more. I added this because when I added email/username uniqueness check into map adapter import test started to fail, because there is LDAP configured, which means when a user was imported, the check was called, and it tried to check for ldap users with the given email, that failed because ldap is not running it just checks whether ldap configuration was imported. 

However, since we changed the check to just check current storage, I think we can remove this.

File: testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/BackwardsCompatibilityUserStorage.java [hunk 6]
```diff
@@ -334,7 +334,7 @@ public int getUsersCount(RealmModel realm) {
     }
 
     @Override
-    public Stream<UserModel> searchForUserStream(String search, RealmModel realm, int firstResult, int maxResults) {
+    public Stream<UserModel> searchForUserStream(String search, RealmModel realm, Integer firstResult, Integer maxResults) {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/BackwardsCompatibilityUserStorage.java:337]: Are you sure? I changed int -> Integer for all stream variant, hence we won't break backward compatibility, because stream variants are new and old implementations won't be affected. However, since all implementations in our code implement XXX.Stream variants of interfaces we should change them. Does it make sense? Or I should change to int even new methods.

File: server-spi/src/main/java/org/keycloak/models/UserModel.java [hunk 7]
```diff
@@ -143,6 +158,11 @@
 
     String getEmail();
 
+    /**
+     * Sets email for this user, email should be switched to lowercase before setting.
```
[the remark this answers is not in the document]
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/models/UserModel.java:162]: This is because JPA layer is doing it [this way](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java#L331). 

If we decide to change this to not use lowercase, we should probably do this in separate Jira across whole codebase.

File: server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java (hunk 1 above)
Comment (mhajas) [reply server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java:30]: WDYT?

```Therefore, each implementation of this interface should consider persisting events and act upon them only when the main transaction was successful. This can be achieved by enlisting the internal transaction to {@link org.keycloak.models.KeycloakTransactionManager} using the {@link org.keycloak.models.KeycloakTransactionManager#enlistAfterCompletion(KeycloakTransaction)} method.```

File: server-spi/src/main/java/org/keycloak/models/UserModel.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply server-spi/src/main/java/org/keycloak/models/UserModel.java:162]: JPA is doing it this way to ensure that letter case is ignored for usernames. However, it should not be mandated to be done on the level of the storage. It should only be mandated that search ignores the letter case.

File: testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/BackwardsCompatibilityUserStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/BackwardsCompatibilityUserStorage.java:337]: Ah, I missed that this is the stream variant, you're right. Sorry for false alarm!

**Comment by mhajas** [comment]
@hmlnarik The failure in the Unit Tests seems to be intermittent and present also in master, it doesn't seem to be caused by my changes. Have you seen this somewhere already?

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+    }
+
+    @Override
+    public void updateConsent(RealmModel realm, String userId, UserConsentModel consent) {
+        LOG.tracef("updateConsent(%s, %s, %s)%s", realm, userId, consent, getShortStackTrace());
+
+        MapUserEntity user = getEntityByIdOrThrow(realm, userId);
+        UserConsentEntity userConsentEntity = user.getUserConsent(consent.getClient().getId());
+        if (userConsentEntity == null) {
+            throw new ModelException("Consent or not found for client [" + consent.getClient().getId() + "] and user [" + userId + "]");
```
Comment (sguilhen) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:261]: Typo? "Consent or not found... "

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 9]
```diff
@@ -0,0 +1,875 @@
...
+        }
+
+        userConsentEntity.setGrantedClientScopesIds(
+                consent.getGrantedClientScopes().stream()
+                        .map(ClientScopeModel::getId)
+                        .collect(Collectors.toSet())
+        );
+
+        userConsentEntity.setLastUpdatedDate(Time.currentTimeMillis());
```
Comment (sguilhen) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:270]: Not sure if I fully understand how all this works but shouldn't this update happen in a transaction?

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java [hunk 10]
```diff
@@ -37,13 +37,17 @@
 import java.util.stream.IntStream;
 import org.hamcrest.Matchers;
 import org.junit.Test;
+import org.keycloak.testsuite.federation.UserMapStorage;
```
Comment (vramik) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java:40]: This import is not necessary here.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java [hunk 11]
```diff
@@ -37,13 +37,17 @@
...
+
 import static org.hamcrest.Matchers.hasItem;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assume.assumeThat;
+import static org.junit.Assume.assumeTrue;
```
Comment (vramik) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java:50]: dtto

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPBinaryAttributesTest.java [hunk 12]
```diff
@@ -156,18 +158,16 @@ public void test03WritableMapper() {
...
+        });
 
-        // Try to update him with some big non-LDAP mapped attribute. It will fail
-        try {
-            joe.getAttributes().put("someOtherPhoto", Arrays.asList(JPEG_PHOTO_BASE64));
-            adminClient.realm("test").users().get(joe.getId()).update(joe);
-            Assert.fail("Not expected to successfully update user");
-        } catch (ClientErrorException cee) {
-            // Expected
-        }
```
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPBinaryAttributesTest.java:167]: If I understand it correctly the test is actually changed to test something else than before, right? Can you please explain to me what has changed?

File: model/map/src/main/java/org/keycloak/models/map/user/UserCredentialEntity.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
+package org.keycloak.models.map.user;
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/user/UserCredentialEntity.java:1]: missing copyright

File: model/map/src/main/java/org/keycloak/models/map/user/UserConsentEntity.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
+package org.keycloak.models.map.user;
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/user/UserConsentEntity.java:1]: missing copyright

File: model/map/src/main/java/org/keycloak/models/map/user/UserFederatedIdentityEntity.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
+package org.keycloak.models.map.user;
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/user/UserFederatedIdentityEntity.java:1]: missing copyright

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java [hunk 16]
```diff
@@ -405,17 +405,17 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
...
      */
     default Stream<UserModel> getRoleMembersStream(RealmModel realm, RoleModel role) {
-        return Stream.empty();
+        return getRoleMembersStream(realm, role, null, null);
     }
 
     /**
      * Search for users that have a specific role with a specific roleId.
      *
+     * @param role
```
Comment (vramik) [inline server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:414]: nit: the javadoc's params are not aligned with actual method contract

File: services/src/main/java/org/keycloak/services/resources/admin/ClientScopeResource.java [hunk 17]
```diff
@@ -101,10 +101,11 @@ public Response update(final ClientScopeRepresentation rep) {
 
         try {
             RepresentationToModel.updateClientScope(rep, clientScope);
+            adminEvent.operation(OperationType.UPDATE).resourcePath(session.getContext().getUri()).representation(rep).success();
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/ClientScopeResource.java:104]: out of curiosity: can you please explain to me why we need to send admin event before commiting the transaction?

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you @mhajas, great work. I have few minor comments.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java [hunk 18]
```diff
@@ -152,18 +162,10 @@ private void removeUser(UserEntity user) {
         String id = user.getId();
         em.createNamedQuery("deleteUserRoleMappingsByUser").setParameter("user", user).executeUpdate();
         em.createNamedQuery("deleteUserGroupMembershipsByUser").setParameter("user", user).executeUpdate();
-        em.createNamedQuery("deleteFederatedIdentityByUser").setParameter("user", user).executeUpdate();
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:153]: Yop you are right, with the second comment. If I understood it correctly the way it was before was causing some exception, that is why there was needed to call `flush()` twice and also `clear()`.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 19]
```diff
@@ -0,0 +1,875 @@
...
+        LOG.tracef("getFederatedIdentitiesStream(%s, %s)%s", realm, user.getId(), getShortStackTrace());
+        return getEntityById(realm, user.getId())
+                .map(AbstractUserEntity::getFederatedIdentities).orElseGet(Stream::empty)
+                .map(UserFederatedIdentityEntity::toModel);
+    }
+
+    @Override
+    public FederatedIdentityModel getFederatedIdentity(UserModel user, String socialProvider, RealmModel realm) {
+        LOG.tracef("getFederatedIdentity(%s, %s, %s)%s", realm, user.getId(), socialProvider, getShortStackTrace());
+        return getRegisteredEntityById(realm, user.getId())
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:202]: I thought this is necessary to have support for storing changes done to the model, however this is not possible as FederatedIdentityModel is not doing any changes to the entity. I will update this and I will also check whole provider for such usages of `getRegisteredEntityById`.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 20]
```diff
@@ -0,0 +1,875 @@
...
+import java.util.stream.Stream;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import static org.keycloak.models.UserModel.EMAIL;
+import static org.keycloak.models.UserModel.EMAIL_VERIFIED;
+import static org.keycloak.models.UserModel.FIRST_NAME;
+import static org.keycloak.models.UserModel.LAST_NAME;
+import static org.keycloak.models.UserModel.USERNAME;
+
+public class MapUserProvider implements UserProvider.Streams, UserCredentialStore {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:56]: done

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java (hunk 8 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:261]: done

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java (hunk 9 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:270]: Well this is the opposite scenario than before. This is where we need the registered entity, otherwise it won't be done in transaction. Nice catch! I changed the method `getEntityByIdOrThrow` to `getRegisteredEntityByIdOrThrow`.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java [hunk 21]
```diff
@@ -810,7 +797,7 @@ public int getUsersCount(Map<String, String> params, RealmModel realm, Set<Strin
     }
 
     @Override
-    public Stream<UserModel> searchForUserStream(Map<String, String> attributes, RealmModel realm, int firstResult, int maxResults) {
+    public Stream<UserModel> searchForUserStream(Map<String, String> attributes, RealmModel realm, Integer firstResult, Integer maxResults) {
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:798]: done.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java [hunk 22]
```diff
@@ -89,11 +90,11 @@
  * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
  */
 public class LoginTest extends AbstractTestRealmKeycloakTest {
-
+
```
[the remark this answers is not in the document]
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:93]: I always thought that when I see this in a commit it is something that was there before, and Intellij is automatically fixing it. I hope it is not always the other way around, that I accidentally added it. Thanks for spotting this I will be checking this in future commits.

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java [hunk 23]
```diff
@@ -214,7 +214,7 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
      * @param firstResult
      * @param maxResults
      * @return
-     * @deprecated Use {@link #searchForUserStream(String, RealmModel, int, int) searchForUserStream} instead.
+     * @deprecated Use {@link #searchForUserStream(String, RealmModel, Integer, Integer) searchForUserStream} instead.
      */
     @Deprecated
     List<UserModel> searchForUser(String search, RealmModel realm, int firstResult, int maxResults);
```
[the remark this answers is not in the document]
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:220]: This is good question, I am not sure. @hmlnarik wanted to leave `BackwardsCompatibilityUserStorage` as is to implement the old variant. Not sure about the others. There are more providers in testsuite that implement old variants. If we decide to make all except for `BackwardsCompatibilityUserStorage` implement `Streams` variants I can make it part of Cleanup PR. @hmlnarik @sguilhen WDYT?

File: testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/events/TestEventsListenerProvider.java [hunk 24]
```diff
@@ -19,30 +19,36 @@
...
     }
 
     @Override
     public void onEvent(AdminEvent event, boolean includeRepresentation) {
-        // Save the copy for case when same AdminEventBuilder is used more times during same transaction to avoid overwriting previously referenced event
-        adminEvents.add(copy(event));
+        tx.addAdminEvent(event, includeRepresentation);
     }
-
+
```
[the remark this answers is not in the document]
Comment (mhajas) [reply testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/events/TestEventsListenerProvider.java:51]: done

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 25]
```diff
@@ -0,0 +1,875 @@
+package org.keycloak.models.map.user;
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:1]: done

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java (hunk 16 above)
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:414]: I will update all javadocs as part of Cleanup task. I believe it will be ready soon after merging this one. See https://issues.redhat.com/browse/KEYCLOAK-15524

File: model/map/src/main/test/java/org/keycloak/models/map/user/AbstractUserEntityCredentialsOrderTest.java [hunk 26]
```diff
@@ -0,0 +1,68 @@
+package org.keycloak.models.map.user;
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/test/java/org/keycloak/models/map/user/AbstractUserEntityCredentialsOrderTest.java:1]: done

File: server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java [hunk 27]
```diff
@@ -18,15 +18,54 @@
 package org.keycloak.events;
```
[the remark this answers is not in the document]
Comment (mhajas) [reply server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java:18]: It is there. Should I update it somehow?

File: server-spi-private/src/main/java/org/keycloak/events/EventListenerTransaction.java [hunk 28]
```diff
@@ -0,0 +1,62 @@
+package org.keycloak.events;
```
[the remark this answers is not in the document]
Comment (mhajas) [reply server-spi-private/src/main/java/org/keycloak/events/EventListenerTransaction.java:1]: done

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java (hunk 10 above)
Comment (mhajas) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java:40]: done

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java (hunk 11 above)
Comment (mhajas) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/UserModelTest.java:50]: done

File: server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java (hunk 27 above)
[the remark this answers is not in the document]
Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java:18]: Sorry I've overlooked this one, please ignore.

File: model/map/src/main/java/org/keycloak/models/map/user/UserCredentialEntity.java (hunk 13 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/UserCredentialEntity.java:1]: done

File: model/map/src/main/java/org/keycloak/models/map/user/UserConsentEntity.java (hunk 14 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/UserConsentEntity.java:1]: done

File: model/map/src/main/java/org/keycloak/models/map/user/UserFederatedIdentityEntity.java (hunk 15 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/UserFederatedIdentityEntity.java:1]: done

File: server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java (hunk 27 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java:18]: np

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPBinaryAttributesTest.java (hunk 12 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPBinaryAttributesTest.java:167]: This test is testing that attribute with name `LDAPConstants.JPEG_PHOTO` is not stored in local storage but in LDAP because of the mapper created on line 147. The original implementation was relying on the fact that the database is not even able to store such a long value > 255 character (probably depending on DB). This means that if storing of  `LDAPConstants.JPEG_PHOTO` attribute passed it was not stored in DB but in LDAP. However, if we try to store attribute `someOtherPhoto` with the same value, it will throw `ClientErrorException` because this time DB will refuse the value.

However, with Map storage we can store such values, therefore this test was not working correctly. I changed it so that we are not testing based on DB refusing the value, but we just check local storage for the attribute not to be present there.

Comment (vramik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPBinaryAttributesTest.java:167]: It makes perfect sense to me now, thank you for explanation.

File: services/src/main/java/org/keycloak/services/resources/admin/ClientScopeResource.java (hunk 17 above)
Comment (mhajas) [reply services/src/main/java/org/keycloak/services/resources/admin/ClientScopeResource.java:104]: I changed whole event processing a little, see javadoc in [EventListenerProvider](https://github.com/keycloak/keycloak/pull/7645/files#diff-0f446f0779afee5927b69598da2aa364be4c7a01b1932636615d8afca8aec33bR26-R43).

TLDR: All events are now fired only after successful transaction not in the moment when `success()` function is called like it was before.

I am not sure what was the exact reason her for moving it before commit, but it makes sense since we send all events in afterTransaction.

**Comment by mhajas** [comment]
Thank you very much @sguilhen and @vramik! I incorporated all your suggestions, could you please have another look?

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:220]: At this moment, either option is open. It would be better to follow @vramik's suggestion if that does not represent a big amount of work at this point. In the opposite case, we can postpone this for either cleanup or even later when removing the interaces.

[the remark this answers is not in the document]
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:220]: I will make it part of https://issues.redhat.com/browse/KEYCLOAK-15524. Does it work for you @vramik @hmlnarik?

[the remark this answers is not in the document]
Comment (vramik) [reply server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:220]: It works for me.

**Review by vramik — APPROVED** [review]
@mhajas thank you

**Review by sguilhen — APPROVED** [review]
Great work @mhajas, thanks for addressing all the points. It looks good to me now!

**Review by hmlnarik — APPROVED** [review]
